README.md
# moncow

Moncow allows you to run files that look like the following example with a key board shprtcut `Cmd+Shift+1`. The resulting JSON will be loaded in a separate editor beside the current.

```js
// moncow
// You can `require("lodash" | "bluebird" | "moment")`
// You must set the module.exports to a function
// which is the only function that will be run.
module.exports = async function main() {
  const { MyDB } = await connect(
    "local", // one environment defined in `moncow.connections`
    ["MyDB"] // one or many databases defined in `moncow.connections`
  );
  // Find all recipes that use broccoli
  return MyDB.collection("recipes")
    .find({ ingredients: "broccoli" })
    .toArray()
}
```

## Caveats
* You file must start with `// moncow` on the first line.
* Any `console.log` statements used within the file will be visible in the "Extensions" section of the "Output" window.

## Keybindings
`Cmd+Shift+1` create a new file with a moncow boilerplate
`Cmd+Shift+2` execute current file and show results

## Commands
`Moncow: Show list` - shows list of all currently active connections
`Moncow: End all` - ends all active connections

## Config
Edit the user settings and create an entry like the following.
```jsonc
"moncow.connections": [
  {
    "environment": "vagrant", // name of environment, used to connect to it
    "connectionConfigs": [
      {
        "url": "mongodb://localhost:27017", // connect to Mongo via this url
        "dbNames": [ // following databases will be available in your files
          "Users",
          "Profiles",
          "Transactions"
        ]
      }
    ]
  }
]
```

You can also connect via tunnels.
```jsonc
"moncow.connections": [
  {
    "environment": "production",
    "connectionConfigs": [
      {
        "url": "users.myapp.com",
        "tunnel": true, // moncow will automatically connect to port 27017 of the url and map it internally to a port and connect to it.
        "dbNames": [ // following databases will be available in your files
          "Users",
          "Profiles"
        ]
      },
      {
        "url": "transactions.myapp.com",
        "tunnel": true,
        "dbNames": [
          "Transactions"
        ]
      },
    ]
  }
]
```

CHANGELOG.md
# Change Log

All notable changes to the "moncow" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/test/runTest.ts
import * as path from "path";
import { runTests } from "vscode-test";

async function main() {
  try {
    // The folder containing the Extension Manifest package.json
    // Passed to `--extensionDevelopmentPath`
    const extensionDevelopmentPath = path.resolve(__dirname, "../../");
    // The path to test runner
    // Passed to --extensionTestsPath
    const extensionTestsPath = path.resolve(__dirname, "./suite/index");
    // Download VS Code, unzip it and run the integration test
    await runTests({ extensionDevelopmentPath, extensionTestsPath });
  } catch (err) {
    console.error("Failed to run tests");
    process.exit(1);
  }
}

main();


src/config.ts
import { values, each, keys } from "lodash";

export const dbNames = {
  ArchiveDB: "ArchiveDB",
  CorporateActionsDB: "CorporateActionsDB",
  OrgSeriesDB: "OrgSeriesDB",
  OrgDynamicAggregateSeriesDB: "OrgDynamicAggregateSeriesDB",
  OrgStaticAggregateSeriesDB: "OrgStaticAggregateSeriesDB",
  QikSayDB: "QikSayDB",
  SasbAsset4DB: "SasbAsset4DB",
  TVLArticlesDB: "TVLArticlesDB",
  TVLSSODB: "TVLSSODB",
  twitterDB: "twitterDB",
  XmanDB: "XmanDB",
  XmanSasbv2ArticlesDB: "XmanSasbv2ArticlesDB",
  DiagnosticArticlesDB: "DiagnosticArticlesDB"
};

const startTunnelPortsFrom = 27020;
const allDbs = values(dbNames);

const series = [
  "OrgSeriesDB",
  "OrgDynamicAggregateSeriesDB",
  "OrgStaticAggregateSeriesDB"
];

const settings = {
  vagrant: [
    {
      url: "mongodb://localhost:27017",
      dbNames: allDbs
    }
  ],
  development: [
    {
      url: "dev-i360-v2-app2.tvl.ai",
      tunnel: true,
      dbNames: allDbs
    }
  ],
  staging: [
    {
      url: "staging-i360-v2-rtd1x.tvl.ai",
      tunnel: true,
      dbNames: ["QikSayDB"]
    },
    {
      url: "staging-i360-v2-abadb1x.tvl.ai",
      tunnel: true,
      dbNames: series
    },
    {
      url: "staging-i360-v2-archivedb1.tvl.ai",
      tunnel: true,
      dbNames: ["ArchiveDB"]
    },
    {
      url: "staging-i360-v2-articlesdb1x.tvl.ai",
      tunnel: true,
      dbNames: ["TVLArticlesDB"]
    }
  ],
  production: [
    {
      url: "prod-i360-v2-rtd1x.tvl.ai",
      tunnel: true,
      dbNames: ["QikSayDB"]
    },
    {
      url: "prod-i360-v2-abadb1x.tvl.ai",
      tunnel: true,
      dbNames: series
    },
    {
      url: "prod-i360-v2-archivedb1.tvl.ai",
      tunnel: true,
      dbNames: ["ArchiveDB"]
    },
    {
      url: "prod-i360-v2-articlesdb1x.tvl.ai",
      tunnel: true,
      dbNames: ["TVLArticlesDB"]
    }
  ]
};

function normalizeSettings(settings: any): any {
  let port = startTunnelPortsFrom;
  const result: any = {};
  each(settings, (urlSettings, env) => {
    const envSettings: any = {};
    each(urlSettings, ({ url, tunnel, dbNames }) => {
      each(dbNames, dbName => {
        envSettings[dbName] = { url, tunnel, dbName, env };
        if (tunnel) {
          envSettings[dbName].port = port;
          port += 1;
        }
      });
    });
    result[env] = envSettings;
  });
  return result;
}

export const environments = keys(settings);
export const databases = normalizeSettings(settings);


src/runner.ts
import * as _ from "lodash";
import { mapSeries } from "bluebird";
import { MongoClient, Db } from "mongodb";
import { spawn } from "child_process";
import { databases } from "./config";

const { logRow } = require("log-row");

const rowSetting = logRow({
  columns: [
    { key: "method", label: null, width: 20, align: "left" },
    { key: "env", label: "Env", width: 8, align: "left" },
    { key: "dbName", label: "DB", width: 27, align: "left" },
    { key: "url", label: "URL", width: 32, align: "left" },
    { key: "mongoUrl", label: "Mongo", width: 32, align: "left" },
    { key: "tunnel", label: "Tunnel", width: 1, align: "left", truncate: true },
    { key: "cached", label: "Cached", width: 1, align: "left", truncate: true },
    { key: "code", label: "Code" },
    { key: "signal", label: "Signal" },
    { key: "message", label: "Message" }
  ]
});

const row = (...args: any) =>
  console.log("{ MonCow } Runner", rowSetting(...args));

const clients = new Map();
const mongoConfig = { useNewUrlParser: true, useUnifiedTopology: true };

async function cleanUp(setting: any) {
  const { url, childAlive } = setting;
  if (clients.has(url)) {
    const { client, tunnelChild } = clients.get(url);
    await client.close();
    if (tunnelChild && childAlive) {
      tunnelChild.kill("SIGHUP");
    }
    clients.delete(url);
    return url;
  }
}

function createTunnel(setting: any) {
  const method = "createTunnel";
  const { url, port } = setting;
  row({ method, ...setting });
  const tunnelChild = spawn("ssh", [
    "-L",
    `${port}:localhost:27017`,
    "-N",
    url
  ]);
  setting.childAlive = true;
  tunnelChild.stderr.on("data", message =>
    row({ method, ...setting, message })
  );
  tunnelChild.on("exit", () => {
    setting.childAlive = false;
  });
  tunnelChild.on("close", async (code, signal) => {
    try {
      row({ method, ...setting, code, signal });
      await cleanUp(setting);
    } catch (error) {
      row({ method, ...setting, message: error.stack });
    }
  });
  return tunnelChild;
}

async function getMongoClient(setting: any) {
  const { url, tunnel, port } = setting;
  if (clients.has(url)) {
    return clients.get(url).client;
  }
  if (!tunnel) {
    const client = await MongoClient.connect(url, mongoConfig);
    clients.set(url, { client });
    return client;
  }
  const tunnelChild = createTunnel(setting);
  const mongoUrl = `mongodb://localhost:${port}`;
  const client = await MongoClient.connect(mongoUrl, mongoConfig);
  clients.set(url, { client, tunnelChild, port });
  return client;
}

async function loadDb(env: string, dbName: string): Promise<Db> {
  const setting = _.get(databases, [env, dbName]);
  if (!setting) {
    throw new Error(`No settings for dbName ${env}.${dbName}`);
  }
  const client = await getMongoClient(setting);
  return client.db(dbName);
}

export async function connect(env: string, dbNames: string[]): Promise<any> {
  const setting = databases[env];
  if (!setting) {
    throw new Error(`No settings for "${env}"`);
  }
  const result: any = {};
  await mapSeries(dbNames, async (dbName: string) => {
    result[dbName] = await loadDb(env, dbName);
  });
  return result;
}

export async function disconnect(env: string): Promise<string[]> {
  const disconnected = await mapSeries(_.toArray(databases[env]), cleanUp);
  return _(disconnected)
    .compact()
    .uniq()
    .value();
}


src/extension.ts
import * as vscode from "vscode";
import * as moncow from "./moncow";
import boilerplate from "./boilerplate";

const maxContentLength: number = 1000000;
const mustStartWith: string = "// moncow";

async function createNewEditor(
  content: string = "",
  language: string = "jsonc",
  position: number = vscode.ViewColumn.Beside
): Promise<vscode.TextEditor> {
  const document: vscode.TextDocument = await vscode.workspace.openTextDocument({
    content,
    language
  });
  const editor: vscode.TextEditor = await vscode.window.showTextDocument(document, position, true);
  return editor;
}

function getMetaData(originEditor: vscode.TextEditor): { sourceFile: string; ranAt: string } {
  return {
    sourceFile: originEditor.document.uri.path,
    ranAt: new Date().toISOString()
  };
}

function truncate(content: string): string {
  if (content.length > maxContentLength) {
    const truncated: string = content.slice(0, maxContentLength);
    return [
      `// File was truncated to ${maxContentLength} from ${content.length} characters.`,
      "// Please consider tightening your query.",
      truncated
    ].join("\n");
  }
  return content;
}

async function runFile(originEditor: vscode.TextEditor): Promise<void> {
  try {
    const code: string = originEditor.document.getText().trim();
    if (!code.startsWith(mustStartWith)) {
      vscode.window.showInformationMessage(
        `File will not be executed unless it starts with ${mustStartWith}`
      );
      return;
    }
    const { sourceFile, ranAt } = getMetaData(originEditor);
    const { environments, result } = await moncow.run(code);
    const data = { sourceFile, ranAt, environments, result };
    await createNewEditor(truncate(JSON.stringify(data, null, "  ")));
  } catch (error) {
    vscode.window.showInformationMessage(`Failed to run file. ${error.message}`);
  }
}

async function updateSettings(event: vscode.ConfigurationChangeEvent): Promise<void> {
  if (event.affectsConfiguration("moncow")) {
    await moncow.initialize();
  }
}

async function showList(): Promise<void> {
  const connected = await moncow.showConnected();
  await createNewEditor(JSON.stringify({ connected }, null, "  "));
}

async function createFile(): Promise<void> {
  await createNewEditor(boilerplate, "javascript", vscode.ViewColumn.Active);
}

async function end(): Promise<void> {
  const result: string[] = await moncow.disconnectAll();
  vscode.window.showInformationMessage(
    `Ended ${result.length} connection(s).\n${result.join("\n")}`
  );
}

export async function activate({ subscriptions }: vscode.ExtensionContext): Promise<void> {
  await moncow.initialize();
  subscriptions.push(vscode.commands.registerTextEditorCommand("moncow.createFile", createFile));
  subscriptions.push(vscode.commands.registerTextEditorCommand("moncow.runFile", runFile));
  subscriptions.push(vscode.commands.registerTextEditorCommand("moncow.list", showList));
  subscriptions.push(vscode.commands.registerTextEditorCommand("moncow.end", end));
  subscriptions.push(vscode.workspace.onDidChangeConfiguration(updateSettings));
}

export function deactivate(): void {}


src/moncow.ts
import * as vscode from "vscode";
import { mapSeries } from "bluebird";
import { MongoClient } from "mongodb";
import { spawn, ChildProcessWithoutNullStreams } from "child_process";

interface Configs {
  startPortFrom: number;
  connections: {
    environment: string;
    connectionConfigs: {
      url: string;
      tunnel: boolean;
      dbNames: string[];
    }[];
  }[];
}

interface DatabaseInfo {
  dbName: string;
  environment: string;
  url: string;
  tunnel?: boolean;
  port?: number;
}

interface ClientInfo extends DatabaseInfo {
  mongoClient?: MongoClient;
  childProcess?: ChildProcessWithoutNullStreams;
  childProcessIsAlive?: boolean;
}

type Environment = Map<string, DatabaseInfo>;
type ConnectionSettings = Map<string, Environment>;

let connectionSettings: ConnectionSettings = new Map();
const clients: Map<string, ClientInfo> = new Map();
const baseMongoUrl: string = "mongodb://localhost";
const baseTunnelUrl: string = "localhost:27017";
const mongoConfig: object = { useNewUrlParser: true, useUnifiedTopology: true };
const killSignal: string = "SIGHUP";

export function updateSettings(): void {
  const configs: Configs | undefined = vscode.workspace.getConfiguration().get("moncow");
  if (!configs) {
    return;
  }
  const { startPortFrom = 27020, connections = [] } = configs;
  let port: number = startPortFrom;
  connectionSettings = connections.reduce((acc, { environment, connectionConfigs }) => {
    const dbSettings: Map<string, DatabaseInfo> = new Map();
    connectionConfigs.forEach(({ url, tunnel, dbNames }) => {
      dbNames.forEach(dbName => {
        const dbInfo: DatabaseInfo = { dbName, environment, url };
        if (tunnel) {
          dbInfo.tunnel = tunnel;
          dbInfo.port = port;
          port += 1;
        }
        dbSettings.set(dbName, dbInfo);
      });
    });
    acc.set(environment, dbSettings);
    return acc;
  }, new Map());
}

async function disconnectOne(clientInfo: ClientInfo): Promise<string> {
  const { url, mongoClient, childProcess, childProcessIsAlive } = clientInfo;
  try {
    if (mongoClient) {
      await mongoClient.close();
    }
    if (childProcess && childProcessIsAlive) {
      childProcess.kill(killSignal);
    }
    clients.delete(url);
    return `${url} (OK)`;
  } catch (error) {
    // TODO: clean up when disconnection fails
    return `${url} (Failed. ${error.message})`;
  }
}

export async function disconnectAll(): Promise<string[]> {
  const clientsToDisconnect: ClientInfo[] = Array.from(clients.values());
  const result: string[] = await mapSeries(clientsToDisconnect, disconnectOne);
  return result;
}

function attachTunnel(clientInfo: ClientInfo): void {
  const { url, port } = clientInfo;
  const commandArgs: string[] = ["-L", `${port}:${baseTunnelUrl}`, "-N", url];
  const childProcess: ChildProcessWithoutNullStreams = spawn("ssh", commandArgs);
  clientInfo.childProcess = childProcess;
  clientInfo.childProcessIsAlive = true;
  childProcess
    .on("exit", () => {
      clientInfo.childProcessIsAlive = false;
    })
    .once("error", () => disconnectOne(clientInfo));
}

function getMongoUrl(clientInfo: ClientInfo): string {
  const { url, port, tunnel } = clientInfo;
  if (tunnel) {
    return `${baseMongoUrl}:${port}`;
  }
  return url;
}

async function createClient(dbInfo: DatabaseInfo): Promise<ClientInfo> {
  const { url, tunnel } = dbInfo;
  const clientInfo: ClientInfo = { ...dbInfo };
  if (tunnel) {
    attachTunnel(clientInfo);
  }
  clientInfo.mongoClient = await MongoClient.connect(getMongoUrl(clientInfo), mongoConfig);
  clients.set(url, clientInfo);
  return clientInfo;
}

export async function connectOne(environment: string, dbNames: string[]): Promise<object> {
  const envSetting: Environment | undefined = connectionSettings.get(environment);
  if (!envSetting) {
    throw new Error(
      `Attempting to connect to environment "${environment}" which is not defined in "moncow.connections".`
    );
  }
  const dbInfos: DatabaseInfo[] = [];
  const allowedDbNames = Array.from(envSetting.keys())
    .map(dbName => `"${dbName}"`)
    .join(", ");
  dbNames.forEach(dbName => {
    const dbInfo: DatabaseInfo | undefined = envSetting.get(dbName);
    if (dbInfo) {
      dbInfos.push(dbInfo);
    } else {
      throw new Error(
        `Attempting to connect to database "${dbName}" which is not defined in "moncow.connections". Allowed databases for environment "${environment}" are ${allowedDbNames}.`
      );
    }
  });
  if (dbInfos.length === 0) {
    throw new Error(
      `None of the provided database names are defined in "moncow.connections". Allowed databases for environment "${environment}" are ${allowedDbNames}`
    );
  }
  const result: any = {};
  await mapSeries(dbInfos, async dbInfo => {
    const { url, dbName } = dbInfo;
    if (!clients.has(url)) {
      await createClient(dbInfo);
    }
    const client: ClientInfo | undefined = clients.get(url);
    if (client && client.mongoClient) {
      result[dbName] = client.mongoClient.db(dbName);
    }
  });
  return result;
}

export async function run(code: string): Promise<{ result: any; environments: string[] }> {
  const innerModule: { exports?: Function } = {};
  const fn: Function = new Function("require", "connect", "module", code);
  const environments: string[] = [];
  fn(
    require,
    (environment: string, dbNames: string[]) => {
      environments.push(environment);
      return connectOne(environment, dbNames);
    },
    innerModule
  );
  const result: any =
    innerModule.exports && typeof innerModule.exports === "function"
      ? await innerModule.exports()
      : { message: "You did not export an async function" };
  return { result, environments };
}

export async function initialize() {
  await disconnectAll();
  updateSettings();
}

export function showConnected(): any {
  const result: any = {};
  clients.forEach(({ environment, dbName, url, tunnel, port, childProcessIsAlive }) => {
    const env = result[environment] || [];
    env.push({ dbName, url, tunnel, port, childProcessIsAlive });
    result[environment] = env;
  });
  return result;
}
